coderd/prebuilds/state.go

```diff
@@ -0,0 +1,200 @@
+// otherwise creates new ones to reach the desired count.
+func (p PresetSnapshot) handleActiveTemplateVersion() (*ReconciliationActions, error) {
+	state := p.CalculateState()
+
+	// If we have more prebuilds than desired, delete the oldest ones
+	if state.Extraneous > 0 {
+		return &ReconciliationActions{
+			ActionType: ActionTypeDelete,
+			DeleteIDs:  p.getOldestPrebuildIDs(int(state.Extraneous)),
+		}, nil
+	}
+
+	// Calculate how many new prebuilds we need to create
+	// We subtract starting prebuilds since they're already being created
+	prebuildsToCreate := max(state.Desired-state.Actual-state.Starting, 0)
+
+	return &ReconciliationActions{
+		ActionType: ActionTypeCreate,
+		Create:     prebuildsToCreate,
+	}, nil
+}
+
+// handleInactiveTemplateVersion deletes all running prebuilds except those already being deleted
+// to avoid duplicate deletion attempts.
+func (p PresetSnapshot) handleInactiveTemplateVersion() (*ReconciliationActions, error) {
+	state := p.CalculateState()
+
+	// TODO(yevhenii): is it correct behavior? What if we choose prebuild IDs that are already being deleted?
+	prebuildsToDelete := max(len(p.Running)-int(state.Deleting), 0)
+	deleteIDs := p.getOldestPrebuildIDs(prebuildsToDelete)
+
+	return &ReconciliationActions{
+		ActionType: ActionTypeDelete,
+		DeleteIDs:  deleteIDs,
+	}, nil
+}
+
+// needsBackoffPeriod checks if we should delay prebuild creation due to recent failures.
+// If there were failures, it calculates a backoff period based on the number of failures
+// and returns true if we're still within that period.
+func (p PresetSnapshot) needsBackoffPeriod(clock quartz.Clock, backoffInterval time.Duration) (*ReconciliationActions, bool) {
+	if p.Backoff == nil || p.Backoff.NumFailed == 0 {
+		return nil, false
+	}
+	backoffUntil := p.Backoff.LastBuildAt.Add(time.Duration(p.Backoff.NumFailed) * backoffInterval)
+	if clock.Now().After(backoffUntil) {
+		return nil, false
+	}
+
+	return &ReconciliationActions{
+		ActionType:   ActionTypeBackoff,
+		BackoffUntil: backoffUntil,
+	}, true
+}
+
+// countEligible returns the number of prebuilds that are ready to be claimed.
+// A prebuild is eligible if it's running and its agents are in ready state.
+func (p PresetSnapshot) countEligible() int32 {
+	var count int32
+	for _, prebuild := range p.Running {
+		if prebuild.Ready {
+			count++
+		}
+	}
+	return count
+}
+
+// countInProgress returns counts of prebuilds in transition states (starting, stopping, deleting).
+// These counts are tracked at the template level, so all presets sharing the same template see the same values.
+func (p PresetSnapshot) countInProgress() (int32, int32, int32) {
+	var starting, stopping, deleting int32
+
+	// In-progress builds are tracked at the template level, not per preset.
+	// This means all presets sharing the same template will see the same counts
+	// for starting, stopping, and deleting prebuilds.
+	// TODO(yevhenii): is it correct behavior?
```

Comment:
Mentioned in another place as well.

Reply:
`CountInProgressPrebuilds` does indeed return this value as a template-level.
Depends on the outcome of https://github.com/coder/coder/pull/17261/files#r2033813934.

Reply:
It was resolved in another thread. Removing this comment, it's no longer correct.